### ISSUE TITLE
Cannot add or update a child row: a foreign key constraint fails

### ISSUE THREAD

**Issue by roberttolton** [putyourlightson/craft-blitz#30]
Hi,

I'm getting a lot of these errors, which I think might be what's bringing down my site (can't be certain, but they are the only loggable error happening at the same time), causing entry saving to time-out, resulting in PHP crashing and an NGINX 502 response.

```
Next yii\db\IntegrityException: SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`db_live`.`craft_blitz_elementcaches`, CONSTRAINT `craft_blitz_elementcaches_cacheId_fk` FOREIGN KEY (`cacheId`) REFERENCES `craft_blitz_caches` (`id`) ON DELETE CASCADE)
The SQL being executed was: INSERT INTO `craft_blitz_elementcaches` (`cacheId`, `elementId`, `dateCreated`, `uid`, `dateUpdated`) VALUES (292168, 1023, '2018-10-29 10:18:51', 'ab2300d6-9e36-407c-86cb-7c92f0d81e98', '2018-10-29 10:18:51') in /home/forge/website.com/vendor/yiisoft/yii2/db/Schema.php:664
```

I've emptied my Blitz cache, which has 45601 pages in it but it didn't help.

If there's anymore information I can give, let me know.

**Comment by roberttolton** [comment]
I tried disabling Blitz, and my problems with saving entries disappeared. Something is causing a timeout, with Blitz enabled during entry save. This site has around ~7000 entries, and when nearly fully cached it has around 45000 items in Blitz cache. But clearing the cache doesn't enable me to save entries, only disabling the plugin does.

The error has been causing Yii to throw a `HeadersAlreadySentException` too.

My `craft_blitz_elementcaches` table has ~3,713,317 rows within it.

Clearing all three of blitz's tables manually is the only way to resolve the issue, until they fill up again.

**Comment by putyourlightson** [comment]
Thanks for reporting this, we're currently working on some optimisations which should help resolve your issues. We should have a release ready by the end of the week.

**Comment by roberttolton** [comment]
Thanks @putyourlightson On a note of that about clearing the right elements in #31 , in my testing of this issue yesterday, I did notice that, for example:

`craft_blitz_elementcaches` had ~40,000 rows. Saving an entry, in 3 locales with 3 matrix blocks each should only need to clear about ~12 elements? (3 post locales, 3 matrix time 3). However, after saving this entry the table dropped down to 66 rows. Saving one entry cleared nearly 40,000 rows in a single query. Think it might be picking up on other elements like navigation nodes and things?

**Comment by putyourlightson** [comment]
Yeah it really depends on the entry that was saved. If it was an entry that is used as a navigation item then it would in theory need to clear every page that has a navigation menu, in case the title or URL was modified, which may well end up being the entire site. As I said though, the next release should contain some optimisations to help manage the size of the tables in the database.

**Comment by putyourlightson** [comment]
We just released 1.6.0. Please let me know if you run into any other issues.
